[flang] Incorrect error handling of bad AT descriptors in FORMAT statement

Consider this good test with AT descriptors:
```
print '(AT,A)', "  string   ", "|"
print 42, "  string   ", "|"
42 format (AT,A)
end 
```
It produces correct output:
```
$ flang trim-good.f90 && ./a.out
  string|
  string|
```
Let's use invalid `AT10` descriptor in the first `print` statement:
```
print '(AT10,A)', "  string   ", "|"
print 42, "  string   ", "|"
42 format (AT,A)
end 
```
This produces expected error message:
```
$ flang trim-fail-1.f90 && ./a.out
error: Semantic errors in trim-fail-1.f90
trim-fail-1.f90:1:11: error: 'AT' edit descriptor does not accept a width value
  print '(AT10,A)', "  string   ", "|"
            ^^
```
Let's then add `AT10` to `format` statement:
```
print '(AT10,A)', "  string   ", "|"
print 42, "  string   ", "|"
42 format (AT10,A)
end 
```
This produces unexpected errors:
```
$ flang trim-fail-2.f90 && ./a.out
error: Could not parse trim-fail-2.f90
trim-fail-2.f90:3:19: error: expected '=>'
  42 format (AT10,A)
                    ^
trim-fail-2.f90:3:4: in the context: pointer assignment statement
  42 format (AT10,A)
     ^
trim-fail-2.f90:3:1: in the context: execution part construct
  42 format (AT10,A)
  ^
trim-fail-2.f90:3:19: error: expected '('
  42 format (AT10,A)
                    ^
trim-fail-2.f90:3:4: in the context: pointer assignment statement
  42 format (AT10,A)
     ^
trim-fail-2.f90:3:1: in the context: execution part construct
  42 format (AT10,A)
  ^
trim-fail-2.f90:3:19: error: expected '='
  42 format (AT10,A)
                    ^
trim-fail-2.f90:3:4: in the context: assignment statement
  42 format (AT10,A)
     ^
trim-fail-2.f90:3:1: in the context: execution part construct
  42 format (AT10,A)
  ^
```
Not only the errors are wrong for the second `print` statement, even the error for the first `print` statement is missing.

flang revision:
```
flang version 23.0.0git (https://github.com/llvm/llvm-project a75e6a59ede46e838e9a02bafe59aa9eb4c1ac60)
Target: x86_64-unknown-linux-gnu
Thread model: posix
```

--- a/flang/include/flang/Common/format.h
+++ b/flang/include/flang/Common/format.h
@@ -735,6 +735,8 @@ template <typename CHAR> bool FormatValidator<CHAR>::Check() {
       }
       if (stmt_ == IoStmtKind::Read) {
         ReportError("'AT' edit descriptor must not be used for input");
+        suppressMessageCascade_ =
+            false; // reset to allow subsequent '.' check to also report
       }
       break;
     case TokenKind::B:

--- a/flang/include/flang/Common/format.h
+++ b/flang/include/flang/Common/format.h
@@ -730,13 +730,13 @@ template <typename CHAR> bool FormatValidator<CHAR>::Check() {
       if (token_.kind() == TokenKind::UnsignedInteger) {
         ReportError("'AT' edit descriptor does not accept a width value");
         NextToken();
-        suppressMessageCascade_ =
-            false; // reset to allow the Read check below to also report
+        // reset to allow the Read check below to also report
+        suppressMessageCascade_ = false;
       }
       if (stmt_ == IoStmtKind::Read) {
         ReportError("'AT' edit descriptor must not be used for input");
-        suppressMessageCascade_ =
-            false; // reset to allow subsequent '.' check to also report
+        // reset to allow subsequent '.' check to also report
+        suppressMessageCascade_ = false;
       }
       break;
     case TokenKind::B:

--- a/flang/lib/Parser/io-parsers.cpp
+++ b/flang/lib/Parser/io-parsers.cpp
@@ -625,7 +625,7 @@ TYPE_PARSER(construct<format::IntrinsicTypeDataEditDesc>(
         mandatoryWidth, noInt, noInt) ||
     construct<format::IntrinsicTypeDataEditDesc>(
         "A " >> ("T " >> pure(format::IntrinsicTypeDataEditDesc::Kind::AT)),
-        noInt, noInt, noInt) ||
+        maybe(width), maybe("." >> digits), noInt) ||
     construct<format::IntrinsicTypeDataEditDesc>(
         "A " >> pure(format::IntrinsicTypeDataEditDesc::Kind::A), maybe(width),
         noInt, noInt) ||

--- a/flang/test/Semantics/io19.f90
+++ b/flang/test/Semantics/io19.f90
@@ -36,13 +36,27 @@
   !ERROR: Unexpected '.' in format expression
   write(*, '(AT10.2)') str
 
-  ! AT with width on input produces two independent errors.
+  ! AT with width on input produces two and three independent errors.
   !ERROR: 'AT' edit descriptor does not accept a width value
   !ERROR: 'AT' edit descriptor must not be used for input
   read(*, '(AT10)') str
+  !ERROR: 'AT' edit descriptor does not accept a width value
+  !ERROR: 'AT' edit descriptor must not be used for input
+  !ERROR: Unexpected '.' in format expression
+  read(*, '(AT10.2)') str
 
   ! FORMAT statements are standalone; the compiler cannot know if they will
   ! be used with READ or WRITE, so no compile-time error is expected here.
 2 format(AT)
   read(*,2) str
+
+  ! AT with width in FORMAT statements; report compiling error, but not
+  ! READ error.
+  !ERROR: 'AT' edit descriptor does not accept a width value
+3 format(AT10)
+  write(*,3) str
+  !ERROR: 'AT' edit descriptor does not accept a width value
+  !ERROR: Unexpected '.' in format expression
+4 format(AT10.2)
+  read(*,4) str
 end

